### DISCUSSION TITLE
define the timezone

### DISCUSSION THREAD

**Discussion by plandolt** [CuyZ/Valinor#447] [Q&A]
Hi there!

I try to map a JSON containing a date string to a `DateTimeImmutable`. During this, I want to set a specific timezone. In regular php code this would be `new DateTimeImmutable('2023-10-24 08:00:00', new DateTimeZone('Europe/Zurich'))`.

How can I achieve that with this library?

**Comment by romm** [comment]
Hi @scuben sorry for the late answer.

The code below should make it possible to set the timezone for a date:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->registerConstructor(
        function (\DateTimeInterface $datetime, \DateTimeZone $timezone): \DateTimeInterface {
            return $datetime->setTimezone($timezone);
        },
    )
    ->mapper()
    ->map(\DateTimeInterface::class, [
        'datetime' => '2024-03-27T21:12:27+00:00',
        'timezone' => 'America/New_York',
    ]);
```

Would it fit your case?

I'll be thinking about adding this to the core.

**Reply by romm** [reply]
@scuben if you have some time, I'd really appreciate if you could take a look at #509!

**Reply by plandolt** [reply]
I am not sure what the source data looked like back in oct 23. But it might be only the datetime string without timezone information. Thus, the question of how to force set the timezone. With your example, I guess I can do the following?

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->registerConstructor(
        function (\DateTimeInterface $datetime): \DateTimeInterface {
            return $datetime->setTimezone(new \DateTimeZone('America/New_York'));
        },
    )
    ->mapper()
    ->map(\DateTimeInterface::class, [
        'datetime' => '2024-03-27T21:12:27+00:00',
    ]);
```

**Reply by romm** [reply]
Hi @scuben I messed up the PR description, it was not supposed to register a constructor for handling timezones. The PR aims to allow the following:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->mapper()
    ->map(\DateTimeInterface::class, [
        'datetime' => '2024-03-28T21:12:27+00:00',
        'timezone' => 'America/New_York',
    ]);
```

And you don't have to register a custom constructor for it.

Tell me what you think of it, I might merge it anytime soon. 🙂
